packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,442 @@
+  }
+  await fs.writeFile(devProfile, JSON.stringify(profileData, null, 4), { encoding: "UTF-8" });
+  const trustDevCertKey = `${PluginNames.LDEBUG}.trustDevCert`;
+  const secrets: Record<string, string> = deserializeDict(await fs.readFile(devUserData, "UTF-8"));
+
+  if (secrets[trustDevCertKey]) {
+    delete secrets[trustDevCertKey];
+    await fs.writeFile(devUserData, serializeDict(secrets), { encoding: "UTF-8" });
+  }
+}
+
+function getConfigDevJson(appName: string): EnvConfig {
+  const envConfig: EnvConfig = {
+    $schema:
+      "https://raw.githubusercontent.com/OfficeDev/TeamsFx/dev/packages/api/src/schemas/envConfig.json",
+    manifest: {
+      description: `You can customize the 'values' object to customize Teams app manifest for different environments. Visit https://aka.ms/teamsfx-config to learn more about this.`,
+      values: {
+        appName: {
+          short: appName,
+          full: `Full name for ${appName}`,
+        },
+      },
+    },
+  };
+  return envConfig;
+}
+
+async function queryProjectStatus(fx: string): Promise<any> {
+  const settings: ProjectSettings = await readJson(path.join(fx, "settings.json"));
+  const solutionSettings: AzureSolutionSettings =
+    settings.solutionSettings as AzureSolutionSettings;
+  const plugins = getActivatedResourcePlugins(solutionSettings);
+  const envDefaultJson: { solution: { provisionSucceeded: boolean } } = await readJson(
+    path.join(fx, "env.default.json")
+  );
+  const hasFrontend = plugins?.some((plugin) => plugin.name === PluginNames.FE);
+  const hasBackend = plugins?.some((plugin) => plugin.name === PluginNames.FUNC);
+  const hasBot = plugins?.some((plugin) => plugin.name === PluginNames.BOT);
+  const hasProvision = envDefaultJson.solution?.provisionSucceeded as boolean;
+  return { hasFrontend, hasBackend, hasBot, hasProvision };
+}
+
+async function getMultiEnvFolders(projectPath: string): Promise<any> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  const fxConfig = path.join(fx, InputConfigsFolderName);
+  const templateAppPackage = path.join(projectPath, "templates", AppPackageFolderName);
+  const fxPublishProfile = path.join(fx, PublishProfilesFolderName);
+  await fs.ensureDir(fxConfig);
+  await fs.ensureDir(templateAppPackage);
+  return { fx, fxConfig, templateAppPackage, fxPublishProfile };
+}
+
+async function removeOldProjectFiles(projectPath: string): Promise<void> {
+  const fx = path.join(projectPath, `.${ConfigFolderName}`);
+  await fs.remove(path.join(fx, "env.default.json"));
+  await fs.remove(path.join(fx, "default.userdata"));
+  await fs.remove(path.join(fx, "settings.json"));
+  await fs.remove(path.join(fx, "local.env"));
+  await fs.remove(path.join(projectPath, AppPackageFolderName));
+  await fs.remove(path.join(fx, "new.env.default.json"));
+  // version <= 2.4.1, rmove .fx/appPackage.
+  await fs.remove(path.join(fx, AppPackageFolderName));
+}
+
+async function ensureProgrammingLanguage(
```

Comment:
nit: update the naming since it also handle the `defaultFunctionName`.

Reply:
Thanks, refactor to `ensureLanguageAndFunctionName()`